staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -47,36 +48,65 @@ type openAPIAggregator struct {
 	// provided for dynamic OpenAPI spec
 	openAPIService *handler.OpenAPIService
 
-	// Aggregator's OpenAPI spec (holds apiregistration group).
-	aggregatorOpenAPISpec *spec.Swagger
+	contextMapper request.RequestContextMapper
+}
 
-	// Local (in process) delegate's OpenAPI spec.
-	inProcessDelegatesOpenAPISpec *spec.Swagger
+func (s *openAPIAggregator) LoadAndAddLocalSpec(localHandler http.Handler, name string, filterPaths []string) error {
+	spec, err := s.downloadOpenAPISpec(localHandler)
+	if err != nil {
+		return err
+	}
+	if filterPaths == nil {
+		filterPaths = []string{"/apis/"}
+	}
+	if len(filterPaths) > 0 {
+		aggregator.FilterSpecByPaths(spec, filterPaths)
+	}
+	s.addLocalSpec(spec, name)
+	return nil
+}
 
-	contextMapper request.RequestContextMapper
+func (s *openAPIAggregator) addLocalSpec(spec *spec.Swagger, name string) {
+	localApiService := apiregistration.APIService{}
+	localApiService.Name = name
+	localApiService.Spec.Service = nil
+	s.openAPISpecs[name] = &openAPISpecInfo{
+		apiService: localApiService,
+		spec:       spec,
+	}
 }
 
-func buildAndRegisterOpenAPIAggregator(delegateHandler http.Handler, webServices []*restful.WebService, config *common.Config, pathHandler common.PathHandler, contextMapper request.RequestContextMapper) (s *openAPIAggregator, err error) {
+func buildAndRegisterOpenAPIAggregator(delegateHandlers []http.Handler, webServices []*restful.WebService,
+	config *common.Config, pathHandler common.PathHandler, contextMapper request.RequestContextMapper) (s *openAPIAggregator, err error) {
+	if len(delegateHandlers) < 1 {
+		return nil, fmt.Errorf("There should be at least one delegateAPIServer in the chain.")
+	}
 	s = &openAPIAggregator{
 		openAPISpecs:  map[string]*openAPISpecInfo{},
 		contextMapper: contextMapper,
 	}
 
-	// Get Local delegate's Spec
-	s.inProcessDelegatesOpenAPISpec, err = s.downloadOpenAPISpec(delegateHandler)
-	if err != nil {
-		return nil, err
+	for i, chain := range delegateHandlers {
+		pathFilter := []string{"/apis/"}
+		// do not filter paths for the first spec. Use it as a base for merging.
+		if i == 0 {
+			pathFilter = []string{}
+		}
+		s.LoadAndAddLocalSpec(chain, fmt.Sprintf(LocalDelegateChainNamePattern, i), pathFilter)
 	}
 
 	// Build Aggregator's spec
-	s.aggregatorOpenAPISpec, err = builder.BuildOpenAPISpec(
+	aggregatorOpenAPISpec, err := builder.BuildOpenAPISpec(
 		webServices, config)
 	if err != nil {
 		return nil, err
 	}
 	// Remove any non-API endpoints from aggregator's spec. aggregatorOpenAPISpec
 	// is the source of truth for all non-api endpoints.
-	aggregator.FilterSpecByPaths(s.aggregatorOpenAPISpec, []string{"/apis/"})
+	aggregator.FilterSpecByPaths(aggregatorOpenAPISpec, []string{"/apis/"})
+
+	// Reserving non-name spec for aggregator's Spec.
+	s.addLocalSpec(aggregatorOpenAPISpec, "")
```

Comment:
technically the aggregator is in-front of the other servers. Shouldn't we start with its spec and add the others afterwards?